### ISSUE TITLE
remove Cgo from Linux

### ISSUE THREAD

**Issue by hajimehoshi** [hajimehoshi/ebiten#2284]
This is basicaly the same as #1162 but for Linux.

Would the ideas in https://github.com/jart/cosmopolitan/issues/303 be a hint?

**Comment by TotallyGamerJet** [comment]
I believe the same technique used for macOS in purego should work for Linux. Especially now that we are using the normal Cgo calling function `runtime.cgocall` instead of `runtime.libcCall` which isn't available on Linux.

**Comment by hajimehoshi** [comment]
For Linux, the problem is that we cannot use a fix path for libdl.

**Comment by TotallyGamerJet** [comment]
I skimmed through the discussion you posted but didn't read why we can't have a fixed path. Isn't it at `usr/lib/libdl.so.2` or just do `libdl.so.2` and just have the dynamic linker find it? Either way, glibc has merged dl into itself according to [this StackOverflow](https://unix.stackexchange.com/questions/700097/unable-to-load-shared-library-libdl-so-or-one-of-its-dependencies) so you could just link to that. You'd probably have to wait until that version of glibc was standard since it quite new.

**Comment by hajimehoshi** [comment]
> I skimmed through the discussion you posted but didn't read why we can't have a fixed path.

Sorry I left the URL but without thinking carefully. I just thought they were trying to solve an issue without depending on libc (at least when compiling). I think the URL has nothing to do with the fixed-path issue.

>  Isn't it at usr/lib/libdl.so.2 or just do libdl.so.2 and just have the dynamic linker find it? Either way, glibc has merged dl into itself according to this StackOverflow so you could just link to that. You'd probably have to wait until that version of glibc was standard since it quite new.

Oh I see. So we will assume the existence of (g)libc in the user's environment, right?

**Comment by TotallyGamerJet** [comment]
> Oh I see. So we will assume the existence of (g)libc in the user's environment, right?

Well it depends on if libdl.so.2 is __truly__ unable to be found by the dynamic linker than I guess so?

**Comment by hajimehoshi** [comment]
> Well it depends on if libdl.so.2 is truly unable to be found by the dynamic linker than I guess so?

OK that makes sense. It is ok if we don't use it when building an Ebitengine app.

**Comment by hajimehoshi** [comment]
Another issue is that we would have to support various architectures not only 386, amd64, arm, and arm64 but also mips, ppc, and so on:

```
$ go tool dist list | grep linux
linux/386
linux/amd64
linux/arm
linux/arm64
linux/loong64
linux/mips
linux/mips64
linux/mips64le
linux/mipsle
linux/ppc64
linux/ppc64le
linux/riscv64
linux/s390x
```

**Comment by TotallyGamerJet** [comment]
I'd think that doing it for SyscallN would be easy but I'm not sure we'd need to do NewCallback which is a lot of extra code

**Comment by hajimehoshi** [comment]
As new architectures will likely be added in the future for Linux, I think having a Cgo version as a fallback would be feasible (but this might be a tough work for you...)

**Comment by TotallyGamerJet** [comment]
As long as we keep Linux limited to just SyscallN it won’t be. That’s pretty easy to emulate in C

**Comment by TotallyGamerJet** [comment]
Yes. That is the inspiration behind [purego](https://github.com/ebitengine/purego/pull/38)

**Comment by hajimehoshi** [comment]
Note to myself: If this issue is resolved, Ebitengine will be able to run on the Go playground (by using `IMAGE:`) lol

**Comment by TotallyGamerJet** [comment]
I had not heard of `IMAGE:` so I looked it up. Very cool! However, I'm not sure purego or some other mechanism would make this possible. Since Ebitengine uses X11/Wayland and OpenGL neither of which exist on the playground right?

**Comment by TotallyGamerJet** [comment]
But the good news is that if we have a general method for calling into C without Cgo then we can just modify or directly use a C software implementation

**Comment by hajimehoshi** [comment]
This sounds an interesting idea!

**Comment by MetalMaxMX** [comment]
Hello! I have a bit of a question here. 
If Ebitengine eventually supports the removal of Cgo, does that means that cross-compiling from one host to another would be possible? (Say, from Linux to Windows? or maybe even the BSDs)

**Comment by hajimehoshi** [comment]
> from Linux to Windows

It's already possible :)

> or maybe even the BSDs

In theory yes, though Ebitengine is not tested well for BSDs.

**Comment by MetalMaxMX** [comment]
Wait, it's already possible to cross-compile from Linux to Windows!? How? Is it documented anywhere? 
As for the BSDs, well, I believe they'll be fine. Though that's just a gut feeling of mine.

**Comment by hajimehoshi** [comment]
> Wait, it's already possible to cross-compile from Linux to Windows!? How? Is it documented anywhere?

This is done basically by `golang.org/x/sys/windows.NewLazySystemDll` to call functions from DLLs. This is already done since Ebitengine 1.9.

**Comment by MetalMaxMX** [comment]
Oh, I see. Didn't know that admittedly, I saw on a much older issue that cross-building wasn't supported so I just assumed that it wasn't built-in, but now I know. Thanks!

**Comment by hajimehoshi** [comment]
Hm? Shouldn't purego work for non-supported architectures with Cgo? Let me check

**Comment by JupiterRider** [comment]
@hajimehoshi What is the current state of this issue?

**Comment by hajimehoshi** [comment]
No progress. IIUC we need more features to PureGo like accepting and returning structs via C. We want to focus on macOS first anyway.

**Comment by TotallyGamerJet** [comment]
Which APIs on Linux need struct support?

**Comment by hajimehoshi** [comment]
I was confused with https://github.com/hajimehoshi/ebiten/issues/1162#issuecomment-1879982922. Maybe there is no such API for Linux/UNIX.
